### PR TITLE
[vscode] Support TerminalExitReason

### PR DESCRIPTION
**Pull request by rschnekenbu**
#### What it does

Adds support for terminal exit status reason and TerminalExitReason enum 

Fixes [#12023](https://github.com/eclipse-theia/theia/issues/12023)

contributed on behalf of ST Microelectronics

#### How to test
This PR adds a reason why the terminal view was closed. This may come from a user action (close buttton), from the process itself that was exited, from an extension or from tool shutdown. 
Here is an extension based on vscode example with additional information on terminal window close event 

[vscode-terminal-api-example-0.0.1.zip](https://github.com/eclipsesource/theia/files/10795468/vscode-terminal-api-example-0.0.1.zip)

1. Testing close on user action
- open _Command Palette..._ 
- select _Terminal API: Create Terminal_ 
- select _Terminal API: Show Terminal_ and select one terminal
- simply close terminal with 'X' button
- notification shall indicate 3 as exit reason (**User** value)

2. Testing close on process exit
- open _Command Palette..._ 
- select _Terminal API: Create Terminal_ 
- select _Terminal API: Show Terminal_ and select one terminal
- type _'exit'_ in the terminal window
- notification shall indicate 2 as exit reason (**Process** value)

3. Testing close by extension
- open _Command Palette..._ 
- select _Terminal API: Create Terminal_ 
- select _Terminal API: Show Terminal_ and select one terminal
- select _Terminal API: Dispose_ and select one terminal
- notification shall indicate 4 as exit reason (**Extension** value)

I could not find a way to reproduce and to check shutdown reason.

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- [x] As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/plugin/src/theia.d.ts [hunk 1]
```diff
@@ -13620,291 +13655,291 @@ export module '@theia/plugin' {
...
-        * Closes the tab group. This makes the tab group object invalid and the tab group
-        * should no longer be used for further actions.
-        * @param tabGroup The tab group to close.
-        * @param preserveFocus When `true` focus will remain in its current position.
-        * @returns A promise that resolves to `true` when all tab groups have been closed.
-        * @stubbed
-        */
-       close(tabGroup: TabGroup | readonly TabGroup[], preserveFocus?: boolean): Thenable<boolean>;
-   }
+    export class TabInputText {
```
Comment (tsmaeder) [inline packages/plugin/src/theia.d.ts:13658]: This seems like an unrelated change. While drive-by fixes may make sense (to enable testing the changes, for example), usually, the community tries to make PR's that have a unity of purpose. Unless there's something I'm missing, I would lose this change.

Comment (tsmaeder) [inline packages/plugin/src/theia.d.ts:13658]: This seems like an unrelated change. While drive-by fixes may make sense (to enable testing the changes, for example), usually, the community tries to make PR's that have a unity of purpose. Unless there's something I'm missing, I would lose this change.

File: packages/terminal/src/browser/terminal-widget-impl.ts [hunk 2]
```diff
@@ -721,7 +733,7 @@ export class TerminalWidgetImpl extends TerminalWidget implements StatefulWidget
             // Close the backend terminal only when explicitly closing the terminal
             // a refresh for example won't close it.
             this.shellTerminalServer.close(this.terminalId);
-            this.exitStatus = { code: undefined };
+            this.exitStatus = { code: undefined, reason: TerminalExitReason.Extension };
```
Comment (tsmaeder) [inline packages/terminal/src/browser/terminal-widget-impl.ts:736]: It's not obvious why the exit reason would necessarily be `extension` here. What's the reasoning?

File: packages/terminal/src/common/base-terminal-protocol.ts [hunk 3]
```diff
@@ -63,11 +63,20 @@ export namespace IBaseTerminalServer {
 export interface IBaseTerminalExitEvent {
     terminalId: number;
 
-    // Exactly one of code and signal will be set.
+    // Exactly one of code, reason and signal will be set.
```
Comment (tsmaeder) [inline packages/terminal/src/common/base-terminal-protocol.ts:66]: Not sure why `code` and `reason` would not be set at the same time.

**Comment by tsmaeder** [comment]
Could you update the main branch on eclipsesource to get the latest version of the API test suite?

**Comment by tsmaeder** [comment]
The code seems to work as advertised.

File: packages/plugin/src/theia.d.ts (hunk 1 above)
Comment (rschnekenbu) [reply packages/plugin/src/theia.d.ts]: Good catch! This should be fixed with new version of the branch updated with latest master, only changes on TerminalExitReason should be visible.

File: packages/terminal/src/browser/terminal-widget-impl.ts (hunk 2 above)
Comment (rschnekenbu) [reply packages/terminal/src/browser/terminal-widget-impl.ts:737]: The exit status is set when closed by user or by the process. This only leaves the option of being closed by extension or by shutdown, which is not monitored in theia AFAIK. In vscode, the terminal service takes care of the shutdown phase, we do not have something similar. So the only valid option I see here is the extension one. 
I added a comment in the new version of the code, and I would be happy to change if you feel something else should be set here.

**Review by tsmaeder — APPROVED** [review]
:lgtm:

**Comment by JonasHelming** [comment]
@tsmaeder Is this an approval?

**Comment by rschnekenbu** [comment]
I will then push the commit to eclipse-theia main repo
